### PR TITLE
Object detection - SAHI

### PR DESCRIPTION
**Pull request by stegmuel**
## What has changed and why?

Added Slicing Aided Hyper Inference (SAHI) for object detection to improve small objects recall at inference.

## How has it been tested?

Manual tests.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [x] Yes
- [ ] Not needed (internal change)

## Did you update the documentation?

- [x] Yes
- [ ] Not needed (internal change without effects for user)

### CODE REVIEW

File: docs/source/object_detection.md [hunk 1]
```diff
@@ -186,6 +186,27 @@ fig.savefig("predictions.png")
...
+Detecting small objects in high-resolution images can be challenging because they may occupy only a few pixels when the image is resized to the model’s input resolution.
+To address this, we support Slicing Aided Hyper Inference (SAHI) allowing the model to make predictions from overlapping tiles of the original image at full resolution and then merge the predictions.
+
+Using tiled inference requires no extra setup:
+
+```python
+import lightly_train
+
+model = lightly_train.load_model("dinov3/vitt16-ltdetr-coco")
+labels, boxes, scores = model.predict_sahi(image="image.jpg").values()
```
Comment (guarin) [inline docs/source/object_detection.md]:
```suggestion
results = model.predict_sahi(image="image.jpg")
results["labels"]   # Class labels, tensor of shape (num_boxes,)
results["bboxes"]   # Bounding boxes in (xmin, ymin, xmax, ymax) absolute pixel
                    # coordinates of the original image. Tensor of shape (num_boxes, 4).
results["scores"]   # Confidence scores, tensor of shape (num_boxes,)
```

Then it has the same format as elsewhere in the docs.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 2]
```diff
@@ -252,13 +253,129 @@ def predict(
         x = transforms_functional.resize(x, self.image_size)
         x = x.unsqueeze(0)
 
-        labels, boxes, scores = self(x, orig_target_size=(h, w))
+        labels, boxes, scores = self(x, orig_target_size=torch.tensor([[h, w]]))
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Add device when initializing the tensor

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 3]
```diff
@@ -252,13 +253,129 @@ def predict(
...
         keep = scores > threshold
         labels, boxes, scores = labels[keep], boxes[keep], scores[keep]
         return {
-            "labels": labels.squeeze(0),
-            "bboxes": boxes.squeeze(0),
-            "scores": scores.squeeze(0),
+            "labels": labels,
+            "bboxes": boxes,
+            "scores": scores,
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:274]: Why is the `squeeze` not needed anymore? There is still `x = x.unsqueeze(0)` on line 254 and the method can only process a single image at the time

File: src/lightly_train/_task_models/object_detection_components/tiling_utils.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+import math
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+from torchvision.ops import box_iou, nms
+
+
+def _tile_image(
```
Comment (guarin) [inline src/lightly_train/_task_models/object_detection_components/tiling_utils.py]:
```suggestion
def tile_image(
```

As function is accessed from outside this file

File: src/lightly_train/_task_models/object_detection_components/tiling_utils.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            )
+
+    # Stack the tiles and coordinates
+    tiles = torch.stack(tiles)
+    tiles_coordinates = torch.stack(tiles_coordinates)
+
+    return tiles, tiles_coordinates
+
+
+def _combine_predictions_tiles_and_global(
```
Comment (guarin) [inline src/lightly_train/_task_models/object_detection_components/tiling_utils.py]:
```suggestion
def combine_predictions_tiles_and_global(
```

As function is accessed from outside this file

**Review by guarin — CHANGES_REQUESTED** [review]

File: docs/source/object_detection.md (hunk 1 above)
Comment (stegmuel) [reply docs/source/object_detection.md]: Should I add the visualization code too? 
It would show that the predictions are in the original image scale.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 3 above)
Comment (stegmuel) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:274]: The `[keep]` flattens everything (that code would not work with batch size > 1 btw).
The squeeze was not needed in the first place I believe.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 6]
```diff
@@ -235,7 +236,7 @@ def load_train_state_dict(
                 if name.startswith("model."):
                     name = name[len("model.") :]
                     new_state_dict[name] = param
-        return self.load_state_dict(new_state_dict, strict=strict, assign=assign)
+        return self.load_state_dict(new_state_dict, strict=False, assign=assign)
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Why is this change needed?

File: docs/source/object_detection.md (hunk 1 above)
Comment (guarin) [reply docs/source/object_detection.md]: I think it is fine without

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 6 above)
Comment (stegmuel) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Reverted.

**Review by guarin — APPROVED** [review]
LGTM
